cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,506 @@
+// Copyright 2021 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	dpapi "github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	podresourcesv1 "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"k8s.io/kubernetes/pkg/kubelet/apis/podresources"
+)
+
+const (
+	// ResourcePrefix is the intel resource prefix.
+	resourcePrefix = namespace + "/"
+
+	gasTSAnnotation   = "gas-ts"
+	gasCardAnnotation = "gas-container-cards"
+
+	grpcAddress    = "unix:///var/lib/kubelet/pod-resources/kubelet.sock"
+	grpcBufferSize = 4 * 1024 * 1024
+	grpcTimeout    = 5 * time.Second
+
+	retryTimeout = 1 * time.Second
+)
+
+// Errors.
+var (
+	errMultiContainer   = errors.New("multi-container allocation request")
+	errMonitoringDevice = errors.New("monitoring device is ignored")
+	errNoAnnotations    = errors.New("annotations missing from container allocate response")
+	errRetry            = errors.New("things didn't work out, but perhaps a retry will help")
+	errZeroPending      = errors.New("there are no pending pods anymore in this node")
+	errBadPath          = errors.New("bad path, major/minor not resolved")
+	errBadCardName      = errors.New("bad card name, major/minor not resolved")
+)
+
+// map of resources. name -> resource amount.
+type resourceMap map[string]int64
+
+type podCandidate struct {
+	pod                 *v1.Pod
+	name                string
+	allocatedContainers int
+	allocationTargetNum int
+}
+
+type resourceManager struct {
```

Comment:
I see `resourceManager` implements the `Scanner` and `PostAllocator` interfaces, but I don't quite understand why. Could you add a doc string here explaining its high level functionality? Unit tests for its API could be explanatory too.

Reply:
I can definitely add a doc string. The reason why unit tests are lacking still, as well as proper documentation, is due to my pessimistic stance towards the amount of changes happening during review. I wouldn't want to write those unit tests twice, ditto for documentation. Meanwhile, you can take a look here: https://github.com/uniemimu/platform-aware-scheduling/tree/gas/gpu-aware-scheduling

The need for Scanner arises from the need to obey the scheduler level card selection and setting the devices to the allocation response accordingly. Basically, the GPU resource manager needs access to the device tree of the GPU plugin.

The task of the PostAllocator here is to take the allocation response, remove the devices and mounts from the response, and replace them with whatever was the decision at the scheduler level. For that, it first finds the POD, and the container, for which the allocation is happening, and it reads the cards from the annotation in the POD, which is written by the GPU Aware Scheduling extender. Then it snatches the desired devices from the devtree and sets them into the allocation response.

In the end then, it writes an annotation for cgroups, but now I hear it is still a moving target so I think I'd better move that cgroup annotation part to another patch outside this PR.

Reply:
Alright. My concern was that such method names suggest the object implements the interfaces used the plugins' gRPC server, but it isn't supposed to be called by the server directly. Perhaps makes sense to rename them to reflect their purpose better.

Reply:
The comment / doc part is still missing from here but I tried to address all your other comments. Removed the cgroups part to another PR, we can forget it for now, and I also removed the operator part to another PR, which can be considered later. I did them renamings etc.